### PR TITLE
[POC/Proposal] Custom output defined by user

### PR DESCRIPTION
**Pull request by shwarcu**
## Intention behind this PR | User story
Allow users to write custom plugin/output/processor (chose your own fancy word :D) in order to work with results returned by Bundlemon.

👨 As an user I would like to write my own integration with bundlemon that would allow to do anything with results - send to custom storage, send slack notification, save in various formats, create a Jira ticket etc.

💡 Inspired by Jest results processor option: https://jestjs.io/docs/configuration#testresultsprocessor-string

### Benefits:
- ⭐ No need to use json output format as a *proxy* format when user needs to do something else with results
- ⭐ Bundlemon more *open* for integrations, it will fits people's needs easier

### Use cases

```json
{
  "verbose": true,
  "baseDir": "./lib/cli",
  "files": [
    {
      "path": "*/*.js"
    },
  ],
  "groups": [
    {
      "path": "**/*.js"
    }
  ],
  "defaultCompression": "none",
  "reportOutput": [["custom", { "path": "custom-output.js" }]]
}

```

Will call function exported by `custom-output.js` with passed report as parameter.
Example file that meets requirements:

```js
const output = (report) => {
  console.log('I AM WORKING');
  console.log(report.status);
};

module.exports = output;

```

### What needs to be done
- support other ways of exporting a function ( I think I lack JS skills for this one, I saw [implementation in Jest](https://github.com/facebook/jest/blob/main/packages/jest-util/src/requireOrImportModule.ts#L13) and I don't understand it yet)
```js
export const output = (report) => {...}
```

```js
const output = (report) => {...}
export default output;
```
- document report object structure
- more testing
- unit tests / integration tests

### CODE REVIEW

File: packages/bundlemon/src/main/outputs/outputs/custom.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+function validateOptions(options: unknown): NormalizedCustomOutputOptions {
+  const schema: yup.SchemaOf<CustomOutputOptions, CustomOutputOptions> = yup.object().required().shape({
+    path: yup.string().required(),
+  });
+
+  const normalizedOptions = validateYup(schema, options, `${NAME} output`);
+  if (normalizedOptions === undefined) {
+    throw new Error(`validation error in output "${NAME}" options`);
+  }
+  return normalizedOptions as NormalizedCustomOutputOptions;
```
Comment (shwarcu) [inline packages/bundlemon/src/main/outputs/outputs/custom.ts:30]: Just for clarity: I hate this hack :sob:

File: packages/bundlemon/src/main/outputs/outputs/custom.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+const output: Output = {
+  name: NAME,
+  create: ({ options }): OutputInstance | Promise<OutputInstance | undefined> | undefined => {
+    const normalizedOptions = validateOptions(options);
+
+    return {
+      generate: async (report: Report): Promise<void> => {
+        const resolvedPath = path.join(process.cwd(), normalizedOptions.path);
```
Comment (shwarcu) [inline packages/bundlemon/src/main/outputs/outputs/custom.ts:40]: Is there a smarter way to make code always refer to the root of the project?

**Comment by LironEr** [comment]
Awesome idea!
I will check the PR later this weekend (:

File: packages/bundlemon/src/main/outputs/outputs/custom.ts (hunk 1 above)
Comment (LironEr) [reply packages/bundlemon/src/main/outputs/outputs/custom.ts:33]: yeah, not sure what do to with that, will try to fix it later

File: packages/bundlemon/src/main/outputs/outputs/custom.ts (hunk 2 above)
Comment (LironEr) [reply packages/bundlemon/src/main/outputs/outputs/custom.ts]:
```
path.resolve(normalizedOptions.path)
```

**Comment by LironEr** [comment]
Sorry for the late response, I'm a bit busy (:

Pushed some changes, let me know what u think.
currently,  `export defaults` will not work, but I'm fine with that for now

**Comment by shwarcu** [comment]
> Sorry for the late response, I'm a bit busy (:
> 
> Pushed some changes, let me know what u think. currently, `export defaults` will not work, but I'm fine with that for now

No problem, I totally understand :)  
I will take a look at the PR around end of the week!

**Comment by shwarcu** [comment]
@LironEr I checked your improvements and also added basic unit tests, would you take a look if they are correct? maybe there is smarter way to test it. Also, when I run unit tests and when they are run by github actions, there are some errors coming from `configUtils `. Probably it needs some mocking or a test key? 
Screenshot 2021-12-13 at 07 51 22

**Comment by LironEr** [comment]
Great! will review soon

Thanks!

**Review by LironEr — APPROVED** [review]

File: docs/output.md [hunk 3]
```diff
@@ -13,6 +13,8 @@
 ]
 ```
 
+---
```
Comment (tcka4enckoalexei2014-commits) [inline docs/output.md:16]: Go

File: docs/output.md [hunk 4]
```diff
@@ -13,6 +13,8 @@
...
+
```
Comment (tcka4enckoalexei2014-commits) [inline docs/output.md:17]: 200

**Pull request merged**